pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,351 @@
+	v signature.Verifier
+}
+
+func (v *verifier) KeyID() (string, error) {
+	return "", nil
+}
+
+func (v *verifier) Public() crypto.PublicKey {
+	// the dsse library uses this to generate a key ID if the KeyID function returns an empty string
+	// as well for the AcceptedKey return value.  Unfortunately since key ids can be arbitrary, we don't
+	// know how to generate a matching id for the key id on the envelope's signature...
+	// dsse verify will skip verifiers whose key id doesn't match the signature's key id, unless it fails
+	// to generate one from the public key... so we trick it by returning nil ¯\_(ツ)_/¯
+	return nil
+}
+
+func (v *verifier) Verify(data, sig []byte) error {
+	if v.v == nil {
+		return errors.New("nil verifier")
+	}
+	return v.v.VerifySignature(bytes.NewReader(sig), bytes.NewReader(data))
+}
+
+func (v V001Entry) CreateFromArtifactProperties(_ context.Context, props types.ArtifactProperties) (models.ProposedEntry, error) {
+	returnVal := models.Dsse{}
+	re := V001Entry{}
+
+	artifactBytes := props.ArtifactBytes
+	if artifactBytes == nil {
+		if props.ArtifactPath == nil {
+			return nil, errors.New("path to artifact file must be specified")
+		}
+		if props.ArtifactPath.IsAbs() {
+			return nil, errors.New("dsse envelopes cannot be fetched over HTTP(S)")
+		}
+
+		var err error
+		artifactBytes, err = ioutil.ReadFile(filepath.Clean(props.ArtifactPath.Path))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	env := dsse.Envelope{}
+	if err := json.Unmarshal(artifactBytes, &env); err != nil {
+		return nil, fmt.Errorf("payload must be a valid dsse envelope: %w", err)
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	if props.PublicKeyBytes != nil {
+		allPubKeyBytes = append(allPubKeyBytes, props.PublicKeyBytes)
+	}
+
+	allPubKeyBytes = append(allPubKeyBytes, props.PublicKeysBytes...)
+	allPubKeyPaths := make([]*url.URL, 0)
+	if props.PublicKeyPath != nil {
+		allPubKeyPaths = append(allPubKeyPaths, props.PublicKeyPath)
+	}
+
+	for _, path := range allPubKeyPaths {
+		if path.IsAbs() {
+			return nil, errors.New("dsse public keys cannot be fetched over HTTP(S)")
+		}
+
+		publicKeyBytes, err := ioutil.ReadFile(filepath.Clean(path.Path))
+		if err != nil {
+			return nil, fmt.Errorf("error reading public key file: %w", err)
+		}
+
+		allPubKeyBytes = append(allPubKeyBytes, publicKeyBytes)
+	}
+
+	keysBySig, err := verifyEnvelope(allPubKeyBytes, &env)
+	if err != nil {
+		return nil, err
+	}
+
+	decodedPayload, err := base64.StdEncoding.DecodeString(env.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode envelope payload: %w", err)
+	}
+
+	paeEncodedPayload := dsse.PAE(env.PayloadType, decodedPayload)
+	h := sha256.Sum256(paeEncodedPayload)
+	re.DsseObj.Payload = strfmt.Base64(env.Payload)
+	re.DsseObj.PayloadType = &env.PayloadType
+	re.DsseObj.PayloadHash = &models.DsseV001SchemaPayloadHash{
+		Algorithm: models.DsseV001SchemaPayloadHashAlgorithmSha256,
+		Value:     hex.EncodeToString(h[:]),
+	}
```

Comment:
this should be recomputed on the server side so no need to compute or set it here.